### PR TITLE
Part 7: Test Results - Cmd and  Report display

### PR DESCRIPTION
**Pull request by lancetipton**
## Context

* The Test Results tab needs some cleanup

## Goal

* Clean up the Test Results tab

## Updates

* Fixed test reports loading
* Fixed issue with backend feature parsing
* Added action to load a test file, based on clicked on test file
  * This then allows loading the latest test results for a test file from the side menu
* Minor cleanup to the cmd output
* Added toggle loading the test-results iframe based on if tests are running
  * Once they are done running, the test-results iframe loads the latest test-results file
* Fixed issue with reports paths
  * This is an odd issue, in the past I needed quotes around the string
    * That does not seem to be the case any longer
    * **This may need to be investigated more**

## Testing
**Setup**
* Move to the keg-herkin local repo and pull down these changes with `keg pr <>`
* Switch to tap-events-force, and pull down this branch => `test-feature-files`
  * Cmd => `keg evf && keg fetch && keg br test-feature-files` 
* While cwd is the tap-events-force repo, start keg-herkin with the following cmd 
  * `keg evf && keg herkin start --from keg-herkin:cmd-report-loading`
* Once it starts navigate to http://herkin-cmd-report-loading.local.kegdev.xyz/ in your browser
* The test settings modal should display
  * Select the `dayNavigation.feature` from the drop down
  * Then click `Do It`
* The feature should load on the page along with a list of all the Definitions

**Test 1**
* Click the Run Tests Button
* This will run the test from the feature
* It should automatically switch to the `Test Results` tab
  * The output from the running test command should display in top `Test Output` section
    * Looks almost like a terminal
  * The `Test Results` Section below it should show a message; `Test Running` with a spinner
  * Once the test finish, the `Test Results Report` should display in the `Test Results` section

> * **IMPORTANT** - At this point it does not matter if the tests fail or not.
> * So don't worry about it when the feature fails
> * If you really care about seeing all green, select the other feature file `Scrollserssions.feature`
>   * Then run the tests on it, that one should pass.
>   * And NO, I don't care that it's spelled `serssions`, that is not the point
  
**Test 2**
* At the bottom of the `Test Results` tab, there should be a `Re-Run Tests` button.
* After running the tests once for a file, click that button, and it should re-run the tests again

**Test 3**
* Doing **Test 1/2** are required before doing this one
* Open the left-hand side menu
* You should see a Reports, section, click on it, then click on the feature sub-section
* Inside the feature sub-section, you'll see `test results` files
* Click on one, and ensure it opens up the test-results section with that feature file

> **NOTES** - It's important to note, that it does not actually open the `results file` you clicked on
> It will open the **most** recent `test results` for the `test file` **related** to the `results file` you clicked on
> But because at this time `feature files` can have **one** test-results file, it ends up being the one you clicked on
> This will change in the future where it will load the actual file you click, but for now, the end result is the same

**Test 4**
* Open the Events-Force Tap in your editor
* Navigate to tests/bdd/steps, and add a new step definition
  * The type and what it does, doesn't matter, it just needs to be valid
  * Something like this is fine => 
  ```js
    const { When } = require('HerkinParkin')
    When('I test a thing', () => {})
  ```
  * Just make sure the matcher string is different then any other definitions
  * Create a new file for it, or put it in an existing definition file, it should not matter
* Once the step definition is created, go back to keg-herkin in the browser
  * In the right side of the `Editor Tab`, you should see the Definitions list
    * On the top right, you will see a new `sync` text and icon
    * Click the sync button, which should cause the definition to reload
    * Ensure the new definition you created, shows up in the list

### CODE REVIEW

**Comment by nnormansv** [comment]
Looks great!  

**Sorting**
In the list of step definitions the group sort is : Given, When, Then
The step expressions appear to be sorted alphabetically within their grouping in the All Steps list however, in the explicit group lists they don't appear to be sorted alpha.
This pull request mentions sorting but didn't specifically say the step expressions would be sorted alpha so may not be related : https://github.com/simpleviewinc/keg-herkin/pull/28

Screen Shot 2021-03-23 at 3 02 21 PM

**Previously Generated Report Cannot Load in UI**
I think is is expected because tests/reports/** are ignored and not committed to the repo, but a previously generated report didn't load in the UI.  It shows in the slide-out as an option but when I clicked on it I got a warning.  The UI handled it gracefully and provided a warning that says, "Could not load related file for /keg/tap/tests/reports/feature/nn-test.feature.html.  It does not exist in the file tree".  It notes the path inside the docker container however the file definitely exists in my local file path at /tap-events-force/tests/reports/feature and there is content inside the file.

How I got into this state : While on a diff branch I created a /tests/bdd/features/nn-test.feature file and ran the test which generated the report in /tests/reports/feature/nn-test.feature.html.    The .feature was not committed.  Then the evf branch was switched for this review and the previously run reports remained.

Screen Shot 2021-03-23 at 2 20 53 PM

Screen Shot 2021-03-23 at 2 22 46 PM

**Review by mpcarolin — CHANGES_REQUESTED** [review]
Looks pretty good, though I encountered one issue that may or may not be related to this PR:

When I started these tests, my `reports/` directory in `tap-events-force` was empty, as expected. Then, when I followed the testing instructions for `Test 3`, I tried to open the tests results using the left-hand side bar like the test instructions say. However, the "Reports" section only said "Empty", even though I had just run the tests (twice). To see the reports listed in the sidebar, I had to first refresh the page, *then* the UI picked up that change.

Also, for `Test 4`, everything worked, but I feel like the UI should just update on its own rather than requiring user to do work. I would prefer if the code just listened for changes in the steps directory and updated itself. Were we to do this, that would probably fix the issue I brought up above.

Anyway, it's all good if these aren't related, but let me know what you think.

I also left a few code comments.

File: tasks/utils/reporter/buildReportTitle.js [hunk 1]
```diff
@@ -11,7 +11,7 @@ const buildReportTitle = (type, context) => {
   const name = context && context.split('/').pop()
   const builtName = name ? wordCaps(name) : type ? `${wordCaps(type)} Test Suite` : `Test Suite`
 
-  return `"${builtName}"`
+  return `${builtName}`
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply tasks/utils/reporter/buildReportTitle.js]: updated

File: tasks/utils/reporter/buildReportPath.js [hunk 2]
```diff
@@ -14,7 +14,7 @@ const buildReportPath = (type, context) => {
   const name = context ? context.split('/').pop() : `${type}s`
   const builtPath = path.join(HERKIN_TESTS_ROOT, HERKIN_REPORTS_DIR, `${type}/${name}.html`)
 
-  return `"${builtPath}"`
+  return `${builtPath}`
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply tasks/utils/reporter/buildReportPath.js]: updated

File: repos/backend/libs/features/featureParser.js [hunk 3]
```diff
@@ -11,7 +11,7 @@ class FeatureParser {
       fs.readFile(location, (err, data) => {
         if(err) return rej(err)
 
-        const parsed = parkin.parse.feature(data.toString())
+        const parsed = (parkin.parse.feature(data.toString())[0] || {})
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/backend/libs/features/featureParser.js]: updated

**Review by mpcarolin — APPROVED** [review]
Looks good, thanks

**Pull request merged**